src/librustc_target/abi/mod.rs

```diff
@@ -913,13 +913,43 @@ pub trait LayoutOf {
     fn layout_of(&self, ty: Self::Ty) -> Self::TyLayout;
 }
 
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum PointerKind {
+    /// Most general case, we know no restrictions to tell LLVM.
+    Shared,
+
+    /// `&T` where `T` contains no `UnsafeCell`, is `noalias` and `readonly`.
+    Frozen,
+
+    /// `&mut T`, when we know `noalias` is safe for LLVM.
+    UniqueBorrowed,
+
+    /// `Box<T>`, unlike `UniqueBorrowed`, it also has `noalias` on returns.
+    UniqueOwned
+}
+
+#[derive(Copy, Clone)]
+pub struct PointeeInfo {
+    pub size: Size,
+    pub align: Align,
+    pub safe: Option<PointerKind>,
+}
+
 pub trait TyLayoutMethods<'a, C: LayoutOf<Ty = Self>>: Sized {
+    type ParamEnv;
+
     fn for_variant(
         this: TyLayout<'a, Self>,
         cx: &C,
         variant_index: VariantIdx,
     ) -> TyLayout<'a, Self>;
     fn field(this: TyLayout<'a, Self>, cx: &C, i: usize) -> C::TyLayout;
+    fn pointee_info_at(
```

Comment:
Each method in this trait needs a "forwarding" wrapper in the `impl` just below, the trait should never be used otherwise, only implemented.

Reply:
Sure, that should be easy to add :+1:. As far as I could tell I am never actually using the trait directly, so that should be alright. Maybe you could give me some more details on the design decision here, so that I can leave it as a comment for the `TyLayoutMethods` trait?

Reply:
Sure. We have to implement the trait on the `Ty` (e.g. `rustc::ty::Ty<'tcx>`), but the functionality should be available as method syntax on `TyLayout<Ty>`.

If we make `TyLayout<Ty>` `Deref` to `Ty`, we can use `self` instead of `this` and have the trait directly callable.
But it already `Deref`s to `LayoutDetails`, sadly.